apex/normalization/instance_norm.py

```diff
@@ -0,0 +1,808 @@
+    InstanceNorm: `stat_axes = [NamedAxis.BATCH, NamedAxis.CHANNEL]`
+
+    Args:
+        fd: An initialized FusionDefinition.
+        x: An input NVFuser tensor.
+        weight: If given, multiply normed output by this `Tensor`. It should be
+            one-dimensional if `NamedAxis.CHANNEL` is in `stat_axes`, and
+            zero-dimensional otherwise. It will be broadcast along all other
+            dimensions.
+        bias: If given, add this `Tensor` to normed output. It should be
+            one-dimensional if `NamedAxis.CHANNEL` is in `stat_axes`, and
+            zero-dimensional otherwise. It will be broadcast along all other
+            dimensions.
+        running_mean: If given, a running mean estimate that will be modified
+            in place.
+        running_var: If given, a running variance estimate that will be
+            modified in place.
+        eps: Amount to regularize the square root needed to convert variance to
+            standard deviation.
+        use_input_stats: Whether to compute the stats of this batch or to
+            _only_ use the provided running_mean and running_var.
+        momentum: Momentum for exponentially weighted moving average of running
+            stats.
+        channels_last: Whether channels are in position -1 (`True`) or 1
+            (`False`).
+        x_datatype: :class:'DataType' of input :class:'Tensor' `x`
+        extent: Size of the input.
+        unbiased: Whether to use unbiased variance for computing current batch
+            statistics. Note that unbiased estimates are always used for
+            running variance updates, regardless of this argument's value.
+        stat_axes: A list of `NamedAxis` objects indicating a combination of
+            axes with which to index the computed statistics. This can be used
+            to implement multiple types of normalization layers, since most of
+            those differ only in which axes are reduced over.
+    Returns:
+        The normalized output, as well as mean and 1/std. Note that
+        `fd.add_output` is _not_ called by this function.
+    """
+    assert not (
+        (running_var is None) ^ (running_mean is None)
+    ), "Iff running mean or var is given, the other should be"
+
+    batch_dim = 0
+    num_dims = len(extent)
+    channel_dim = num_dims - 1 if channels_last else 1
+
+    stat_dims = []
+    # Running stats will be kept possibly for channel but never by instance, so
+    # we will reduce along batch_dim before updating running stats.
+    stat_dims_nobatch = []
+    num_stats = 1
+    if NamedAxis.BATCH in stat_axes:
+        stat_dims.append(batch_dim)
+        num_stats *= extent[batch_dim]
+    if NamedAxis.CHANNEL in stat_axes:
+        stat_dims.append(channel_dim)
+        stat_dims_nobatch.append(channel_dim)
+        num_stats *= extent[channel_dim]
+    x_reduction_axes = [ax for ax in range(num_dims) if ax not in stat_dims]
+    num_features = extent.numel() // num_stats
```

Comment:
Looks like all shape specific things are hard coded as compile time constant.

Just want to point it out here that we are not supporting dynamic shapes? I'm wondering if there's a technical reason that we are not doing so.

Reply:
One issue with dynamic shapes was that unlike in C++, I wasn't sure how to refer to tensor axes/extents in the frontend. @kevinstephano maybe I could take a crack at adding that to the frontend?

Reply:
Pushed 716fbc2e419a8ca307016e7ed8fa29995fb82a3e which enables dynamic shapes: sending any sized tensor does not change the number of cached fusions, even without the explicit cache. The tests are updated to check correctness and that the cache doesn't grow when passing different size inputs.

Reply:
For the record: The trick is to use `fd.ops.tensor_sizes` to get a tuple of `Scalar`s describing the extent of a `Tensor`.